Sonic Suggests v1.2 - Speedrun Edition

Refragg opened the pull request:
Hello! After messing around with the game's code and everything, here I come with an update that I think is best described as the Speedrun Edition, the goal of this update is to keep the game / gameplay itself intact but add features for speedrunners.

That said, this update adds the ability to skip the various cutscenes, a reset button and it also now exposes it's global state externally to allow for autosplitting / load removal

I also went and fixed the ForestAmbiant sound which was for some reason not loading properly in my Unity Editor.

And that's actually really just it, it'd make us really happy if you approved of this update and then we can release it for the speedrun community to have fun with and hopefully, get the game on http://speedrun.com at some point in the future.

With that, I hope you have a nice day!

Refragg committed:
Fix ForestAmbient sound import issues

For some reason, this sound file is using a weird sample rate and stuff and Unity would not be able to import it when loading the project. I just exported it again with standard sample rate and others and it does the trick

Refragg committed:
Add reset game and cutscene skip

This patch adds a cutscene skip feature when you press A in a cutscene and then press it again to confirm, this is implemented for the 3 cutscenes present in the game.
In the same way, it's now possible to reset back to the main menu by pressing the Back / Select button and pressing it again to confirm (this only works when you are already in the game)

Refragg committed:
Expose global game state to the memory

This patch adds a global game state that will be easily accessible from the outside by scanning for the game state signature and then using the known field offsets.
This will notably be used for autospltting and maye other tools if needed.
The way this is done does mean we are using unsafe code but it was the best way i thought of doing this.
Obviously some things can be improved about it, maybe adding some other fields or using properties to avoid duplicate code and stuff, this works for now though and I didn't want to break the Unity part of things that i'm still pretty unfamiliar with.

Refragg committed:
GameState: Change FrameCounter to IsLoading

After some testing and thinking, it seems like a frame counter wouldn't be the best choice for the state so i've decided to change it to a loadless kind of thing

Refragg committed:
GameState: Handle pointer length errors and support 64 bits pointers

This patch adds support for 64 bits (long pointers), The GameState will
detect the used pointer type on the system and use either 32 bits or 64
bits pointers. This fixes the x64 Windows build of the game.
Additionally, it will now handle errors if the pointer length is not
supported to allow the game to continue running.

Refragg committed:
Bump version to 1.2

3DI70R commented:
Hello, i don't mind pull requests, but i'm concerned about GameState implementation. 
It is very unnatural for C# to have pointers. 
I understand that this logic is probably exposes values for autosplitter and they should exist in specific order, preceeding by some signature, but i think you should rewrite it using `[StructLayout(LayoutKind.Explicit)], [FieldOffset(x)]` and fixed size buffers.

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code

That way, game state will be placed in memory in predictable manner, and there will be no pointer arithmetic.

Refragg commented:
Thank you for the review, I will look at writing this in a safer way then, you're right that this might be not the best way of doing this. I actually didn't know about the "Explicit" StructLayout so that'll be very useful for this, I will see when i can do this although I'm somewhat busy right now.

Refragg committed:
GameState: Refactor to not use unsafe code

This patch is a rework of the raw game state in memory and it makes use of non-unsafe patterns and notably, gets rid of raw pointers completly.
This also means that unsafe code is no longer needed to build the game.

Refragg commented:
Alright, i've went and rewrote the GameState to safe code now, what do you say?

3DI70R commented:
Yup, that looks better. 
I'll open this branch in editor and review other changes during this week. 

Though i'm unsure if i release it right away. Earlier this year i've had compulsion that i should make remaster of this game and for whole month i did various improvements for it, but got distracted on another projects. Maybe i should return and finish what i started, and then incorporate your changes in this version. I'll think about it.

![image](https://github.com/3DI70R/Sonic-Suggests/assets/743194/dda6d587-542b-4dd6-a753-988dc124285f)

Refragg commented:
Oh I see, that definitely looks cool! 👀 Good luck if you end up coming back to this then.
Do you allow me to release this version on my fork for now then?

3DI70R commented:
Sure

3DI70R commented:
Sorry if it looks like i've abandoned or ignored your PR
But remaster gradually overgrown its initial scope, and currently its more like remake or something. I've missed Sage 2023 deadline, and went full upgrade, with completely rewritten game core, new enemies, expanded level and other stuff. It's kinda entered "feature creep" phase, but at the moment, its enjoyable one.

"Speedrun mode" is still planned, and i've even implemented ghost recording/playback
https://github.com/3DI70R/Sonic-Suggests/assets/743194/f8a387ea-5384-4945-be02-3f85c606704b

But "1.2" update will take more time than i thought. My current deadline is Sage 2024, it will be around september, so rougly 10 months from now.
